pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/auth/KubernetesServiceAccountAuthProvider.java

```diff
@@ -0,0 +1,160 @@
+
+    private static final String SERVICE_ACCOUNT_VOLUME_NAME = "service-account-token";
+    private static final String TRUST_CERT_VOLUME_NAME = "ca-cert";
+    private static final String DEFAULT_MOUNT_DIR = "/etc/auth";
+    private static final String FUNCTION_AUTH_TOKEN = "token";
+    private static final String FUNCTION_CA_CERT = "ca.crt";
+    private String brokerTrustCertsSecretName;
+    private long serviceAccountTokenExpirationSeconds;
+
+    @Override
+    public void initialize(CoreV1Api coreClient, byte[] caBytes,
+                           java.util.function.Function<Function.FunctionDetails, String> namespaceCustomizerFunc,
+                           KubernetesRuntimeFactoryConfig config) {
+        setNamespaceProviderFunc(namespaceCustomizerFunc);
+        brokerTrustCertsSecretName = config.getBrokerClientTrustCertsSecretName();
+        serviceAccountTokenExpirationSeconds = config.getServiceAccountTokenExpirationSeconds();
+    }
+
+    @Override
+    public void configureAuthenticationConfig(AuthenticationConfig authConfig,
+                                              Optional<FunctionAuthData> functionAuthData) {
+        authConfig.setClientAuthenticationPlugin(AuthenticationToken.class.getName());
+        authConfig.setClientAuthenticationParameters(Paths.get(DEFAULT_MOUNT_DIR, FUNCTION_AUTH_TOKEN)
+                .toUri().toString());
+        // if we have ca bytes, update the new path for the CA
+        if (StringUtil.isNotBlank(brokerTrustCertsSecretName)) {
+            authConfig.setTlsTrustCertsFilePath(String.format("%s/%s", DEFAULT_MOUNT_DIR, FUNCTION_CA_CERT));
+        }
+    }
+
+    /**
+     * No need to cache anything because we rely on Kubernetes to generate the token used for authentication.
+     */
+    @Override
+    public Optional<FunctionAuthData> cacheAuthData(Function.FunctionDetails funcDetails,
+                                                    AuthenticationDataSource authenticationDataSource)
+            throws Exception {
+        return Optional.empty();
+    }
+
+    /**
+     * No need to update anything because we rely on Kubernetes to generate the token used for authentication.
+     */
+    @Override
+    public Optional<FunctionAuthData> updateAuthData(Function.FunctionDetails funcDetails,
+                                                     Optional<FunctionAuthData> existingFunctionAuthData,
+                                                     AuthenticationDataSource authenticationDataSource)
+            throws Exception {
+        return Optional.empty();
+    }
+
+    /**
+     * No need to clean up anything because we rely on Kubernetes to clean up the secret when the pod is deleted.
+     */
+    @Override
+    public void cleanUpAuthData(Function.FunctionDetails funcDetails, Optional<FunctionAuthData> functionAuthData)
+            throws Exception {
+
+    }
+
+    @Override
+    public void initialize(CoreV1Api coreClient) {
```

Comment:
this method is not meant to be called, I suggest to throw an exception here

Reply:
The Pulsar Function Worker does not call this method. We could probably deprecate now and remove it later, if we want. It is used in another implementation here:

https://github.com/apache/pulsar/blob/01a49929951f70f3858a59a15f31cf15c16c1c8e/pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/auth/KubernetesFunctionAuthProvider.java#L40-L45

From an API design perspective, it seems like we only need the `initialize` method and could deprecate the `setCaBytes` and `setNamespaceProviderFunc` methods as well since they are only called in the `initialize` method and never outside of the class.